### ISSUE TITLE
Gnocchi heavily hitting keystone

### ISSUE THREAD

**Issue by dvenza** [gnocchixyz/gnocchi#924]
I configured ceilometer and gnocchi according to the installation guides for Pike on Ubuntu.
I have four compute hosts with polling set to 10 seconds. The problem is that Gnocchi is overwhelming my Keystone (12 processes running under uwsgi) with an extreme amount of requests for the `/v3/auth/tokens` endpoint with requests coming in almost every few milliseconds.

I haven't found a clear way to enable the token cache, I set the memcached_server option, but it seems clear that it is not working.

I can try to scale keystone, but if I add a production-level number of compute hosts I will need a whole datacenter dedicated to keystone?
I found that in Ocata there was a way to disable keystone authentication for Gnocchi in ceilometer, but in Pike seems to have disappeared without any mention.

### Which version of Gnocchi are you using
```
pip freeze | grep gnocchi
gnocchi==4.2.5
gnocchiclient==3.3.1
```
### How to reproduce your problem
Openstack Pike with 4 compute nodes and ~15 VMs running
Ceilometer polling set to 10 seconds

Check CPU usage of the keystone processes.

### What is the result that you get
All Keystone processes jump to 100% CPU usage and stay there, filling the logs with:
```
2018-06-22 06:25:24.573 15888 INFO keystone.common.wsgi [req-6556698c-e441-44a4-9e88-b96fd7797451 d2c0baf8a84647879878c174dc6344cd 80ed43922b204815b6b4186126f7b599 - default default] GET http://bfeb:35357/v3/auth/tokens
2018-06-22 06:25:24.574 15887 INFO keystone.common.wsgi [req-b2b1be17-8c16-49fa-a644-0dab832bbef4 d2c0baf8a84647879878c174dc6344cd 80ed43922b204815b6b4186126f7b599 - default default] GET http://bfeb:35357/v3/auth/tokens
```

(d2c0baf8a84647879878c174dc6344cd  is the ID of the gnocchi user)

### What is result that you expected
Normal Keystone CPU usage.

**Comment by jd** [comment]
This is more likely a question for the Keystone authtoken middleware maintainer. Gnocchi does nothing Keystone specific.

**Comment by dvenza** [comment]
My question is very gnocchi-related:

is it normal for Gnocchi to call keystone with this frequency at this load ?

**Comment by jd** [comment]
Of course, Keystone auth token is called before each and every HTTP request you do on Gnocchi. LIke any other services in OpenStack. There's nothing different.

Then your Keystone auth token middleware configuration might not be state of the art; you could use caching or fernet tokens.

If Ceilometer is configured for doing a lot of polling and pushes a lot of data with a lot of requests, that also can explain why you see so many request to Gnocchi. Also note that you're using a pretty old version of Ceilometer and they might be improvements in the more recent versions related to that.

**Comment by dvenza** [comment]
I'm using both caching and fernet tokens.
So everytime ceilometer (ceilometer-agent-notification) has data to give to gnocchi, it opens an http connection and causes gnocchi to do a token validation, right ?

My ceilometer is from Pike, so I thought not so old as you imply. Looking through the Queens release notes for ceilometer I do not see anything about reducing authentication load on gnocchi.

So basically your answer is that the only places that I should check to understand this load on keystone is either the keystone client (ceilometer and gnocchi side) configuration or the ceilometer code. There is nothing in gnocchi that I can touch to lessen the load. Is this right?

I hope you understand the reasoning behind my question: keystone logs are full of gnocchi calls, not ceilometer.

**Comment by cristicalin** [comment]
@dvenza I recently had the same issue, turns out my keystone auth section in gnocchi had a typo and gnocchi was not able to connect to memcache so instead it was always hitting keystone for token validation.

The keystoneauth middleware does token validation which is what @jd was implying that gnocchi always does the same call but the implementation is up to keystoneauth middleware and it can choose to lookup the token in the cache or forward the validation to keystone in which case it will overload keystone rapidly due mainly to the way ceilometer calls onto gnocchi.

**Comment by dvenza** [comment]
Here is mine and, altough not identical, should be ok:
```ini
[keystone_authtoken]
region_name = regionOne
memcached_servers = localhost:11211
service_token_roles_required = true
auth_type = password
auth_url = http://bfeb:5000                                                                                                                                                     
auth_uri = http://bfeb:5000                                                                                                                                                     
project_domain_name = default                                                                                                                                                   
user_domain_name = default                                                                                                                                                      
project_name = service                                                                                                                                                      
username = gnocchi                                                                                                                                                             
password = xxx
```

memcached is there and is used by other openstack services running on the same host.

Thank you, I'll keep investigating.

**Comment by tsv1991** [comment]
@dvenza hi!
I have same problem! And I can't to solve it. Do you have solving for it?
